File: go/parser/parseContent.go
package parser

import (
	"regexp"
	"sort"
	"strings"

	"github.com/nbd-wtf/go-nostr/nip19"
)

// ContentBlock represents a parsed block of content
type ContentBlock struct {
	Type string                 `json:"type"`
	Text string                 `json:"text"`
	Data map[string]interface{} `json:"data,omitempty"`
}

// Pattern defines a regex pattern and a processor function for content parsing
type Pattern struct {
	Type         string
	Regex        *regexp.Regexp
	ProcessMatch func([]string) (ContentBlock, error)
}

// Match represents a matched pattern in the content with position info
type Match struct {
	Start int
	End   int
	Block ContentBlock
}

// GetLinkPreview is a placeholder for the link preview functionality
// You can implement this later using an appropriate Go library
func GetLinkPreview(url string) (map[string]interface{}, error) {
	// Placeholder - return minimal preview data
	return map[string]interface{}{
		"url":         url,
		"title":       "Link Preview",
		"description": "Link preview not implemented",
	}, nil
}

// ParseContent parses a string content into ContentBlocks
func ParseContent(content string) ([]ContentBlock, error) {
	// Define patterns to match
	patterns := []Pattern{
		{
			Type:  "code",
			Regex: regexp.MustCompile("```([\\s\\S]*?)```"),
			ProcessMatch: func(match []string) (ContentBlock, error) {
				return ContentBlock{
					Type: "code",
					Text: match[0],
					Data: map[string]interface{}{
						"code": match[1],
					},
				}, nil
			},
		},
		{
			Type:  "cashu",
			Regex: regexp.MustCompile("(cashuA[A-Za-z0-9_-]+)"),
			ProcessMatch: func(match []string) (ContentBlock, error) {
				return ContentBlock{
					Type: "cashu",
					Text: match[0],
					Data: map[string]interface{}{
						"token": match[0],
					},
				}, nil
			},
		},
		{
			Type:  "hashtag",
			Regex: regexp.MustCompile(`(?<![^\s"'(])(#[a-zA-Z0-9_]+)(?![a-zA-Z0-9_])`),
			ProcessMatch: func(match []string) (ContentBlock, error) {
				return ContentBlock{
					Type: "hashtag",
					Text: match[0],
					Data: map[string]interface{}{
						"tag": match[0][1:], // Remove the # symbol
					},
				}, nil
			},
		},
		{
			Type:  "image",
			Regex: regexp.MustCompile(`(?i)(https?://\S+\.(jpg|jpeg|png|gif|webp|svg|ico)(\?\S*)?)`),
			ProcessMatch: func(match []string) (ContentBlock, error) {
				return ContentBlock{
					Type: "image",
					Text: match[0],
					Data: map[string]interface{}{
						"src": match[0],
					},
				}, nil
			},
		},
		{
			Type:  "video",
			Regex: regexp.MustCompile(`(?i)(https?://\S+\.(mp4|mov|avi|mkv|webm|m4v)(\?\S*)?)`),
			ProcessMatch: func(match []string) (ContentBlock, error) {
				return ContentBlock{
					Type: "video",
					Text: match[0],
					Data: map[string]interface{}{
						"src": match[0],
					},
				}, nil
			},
		},
		{
			Type:  "nostr",
			Regex: regexp.MustCompile(`(?i)nostr:([a-z0-9]+)`),
			ProcessMatch: func(match []string) (ContentBlock, error) {
				entity := match[1]

				prefix, data, err := nip19.Decode(entity)
				if err != nil {
					// If we can't decode, treat as text
					return ContentBlock{
						Type: "text",
						Text: match[0],
					}, nil
				}

				return ContentBlock{
					Type: prefix, // Will be "npub", "nprofile", "note", "nevent", "naddr"
					Text: match[0],
					Data: map[string]interface{}{
						"decoded": data,
						"bech32":  entity,
					},
				}, nil
			},
		},
		{
			Type:  "link",
			Regex: regexp.MustCompile(`(?i)(https?://\S+)(?!\[\)])`),
			ProcessMatch: func(match []string) (ContentBlock, error) {
				url := match[0]
				if !strings.HasPrefix(strings.ToLower(url), "http") {
					url = "https://" + url
				}

				// Placeholder for link preview
				preview, err := GetLinkPreview("https://proxy.nuts.cash/?url=" + url)
				if err != nil {
					preview = map[string]interface{}{}
				}

				return ContentBlock{
					Type: "link",
					Text: match[0],
					Data: map[string]interface{}{
						"href":    match[0],
						"preview": preview,
					},
				}, nil
			},
		},
	}

	// Find all matches with their positions
	var allMatches []Match

	for _, pattern := range patterns {
		// Find all matches for this pattern
		matches := pattern.Regex.FindAllStringSubmatchIndex(content, -1)

		for _, matchIndices := range matches {
			start := matchIndices[0]
			end := matchIndices[1]

			// Extract the matched text and subgroups
			matchText := content[start:end]
			submatches := []string{matchText}

			// Extract capture groups
			for i := 1; i < len(matchIndices)/2; i++ {
				if matchIndices[2*i] != -1 {
					submatches = append(submatches, content[matchIndices[2*i]:matchIndices[2*i+1]])
				} else {
					submatches = append(submatches, "")
				}
			}

			// Process the match
			block, err := pattern.ProcessMatch(submatches)
			if err != nil {
				continue
			}

			allMatches = append(allMatches, Match{
				Start: start,
				End:   end,
				Block: block,
			})
		}
	}

	// Sort matches by start position
	sort.Slice(allMatches, func(i, j int) bool {
		return allMatches[i].Start < allMatches[j].Start
	})

	// Remove overlapping matches (prioritize earlier patterns)
	var filteredMatches []Match

	for _, match := range allMatches {
		overlaps := false

		for _, existing := range filteredMatches {
			if (match.Start >= existing.Start && match.Start < existing.End) ||
				(match.End > existing.Start && match.End <= existing.End) ||
				(match.Start <= existing.Start && match.End >= existing.End) {
				overlaps = true
				break
			}
		}

		if !overlaps {
			filteredMatches = append(filteredMatches, match)
		}
	}

	// Re-sort filtered matches
	sort.Slice(filteredMatches, func(i, j int) bool {
		return filteredMatches[i].Start < filteredMatches[j].Start
	})

	// Build the final result, including text between matches
	var blocks []ContentBlock
	lastIndex := 0

	for _, match := range filteredMatches {
		// Add text before this match
		if match.Start > lastIndex {
			blocks = append(blocks, ContentBlock{
				Type: "text",
				Text: content[lastIndex:match.Start],
			})
		}

		// Add the match
		blocks = append(blocks, match.Block)

		lastIndex = match.End
	}

	// Add any remaining text after the last match
	if lastIndex < len(content) {
		blocks = append(blocks, ContentBlock{
			Type: "text",
			Text: content[lastIndex:],
		})
	}

	// Post-processing: group consecutive media into grids
	var processedBlocks []ContentBlock
	var mediaGroup []ContentBlock

	for i := 0; i < len(blocks); i++ {
		block := blocks[i]

		// If this is an image or video
		if block.Type == "image" || block.Type == "video" {
			mediaGroup = append(mediaGroup, block)
			continue
		}

		// If this is whitespace or newlines between media, check what follows
		if block.Type == "text" {
			isWhitespace := true
			for _, r := range block.Text {
				if !strings.ContainsRune(" \t\n\r", r) {
					isWhitespace = false
					break
				}
			}

			if isWhitespace {
				// If we have media before and media after, continue collecting
				if len(mediaGroup) > 0 && i+1 < len(blocks) {
					nextBlock := blocks[i+1]
					if nextBlock.Type == "image" || nextBlock.Type == "video" {
						continue
					}
				}
			}
		}

		// If we have collected media and the current block breaks the sequence
		if len(mediaGroup) > 0 {
			// Add media group if it contains more than one item
			if len(mediaGroup) > 1 {
				mediaTexts := make([]string, len(mediaGroup))
				mediaItems := make([]map[string]interface{}, len(mediaGroup))

				for j, media := range mediaGroup {
					mediaTexts[j] = media.Text
					mediaItems[j] = map[string]interface{}{
						"type": media.Type,
						"src":  media.Data["src"],
					}
				}

				processedBlocks = append(processedBlocks, ContentBlock{
					Type: "mediaGrid",
					Text: strings.Join(mediaTexts, "\n"),
					Data: map[string]interface{}{
						"items": mediaItems,
					},
				})
			} else {
				// Just add the single media item
				processedBlocks = append(processedBlocks, mediaGroup[0])
			}
			mediaGroup = nil
		}

		// Add the current non-media block
		processedBlocks = append(processedBlocks, block)
	}

	// Don't forget any remaining media
	if len(mediaGroup) > 0 {
		if len(mediaGroup) > 1 {
			mediaTexts := make([]string, len(mediaGroup))
			mediaItems := make([]map[string]interface{}, len(mediaGroup))

			for j, media := range mediaGroup {
				mediaTexts[j] = media.Text
				mediaItems[j] = map[string]interface{}{
					"type": media.Type,
					"src":  media.Data["src"],
				}
			}

			processedBlocks = append(processedBlocks, ContentBlock{
				Type: "mediaGrid",
				Text: strings.Join(mediaTexts, "\n"),
				Data: map[string]interface{}{
					"items": mediaItems,
				},
			})
		} else {
			processedBlocks = append(processedBlocks, mediaGroup[0])
		}
	}

	return processedBlocks, nil
}

File: src/lib/parseContent.ts
import { nip19, nip27, type AddressPointer, type EventPointer, type ProfilePointer } from 'nostr-tools';
import { marked } from 'marked';

export type ContentBlock = {
	type:
		| 'text'
		| 'image'
		| 'video'
		| 'mediaGrid'
		| 'code'
		| 'link'
		| 'npub'
		| 'nprofile'
		| 'note'
		| 'nevent'
		| 'naddr'
		| 'hashtag'
		| 'cashu';
	text: string;
	data?: Record<string, any>;
};

// Placeholder for markdown links to protect them from NIP-27 URL extraction
const MARKDOWN_LINK_PLACEHOLDER = '\x00LINK\x00';

/**
 * Parse content using NIP-27 to extract references, URLs, images, videos
 * Then apply additional parsing for code blocks, cashu, hashtags
 * 
 * This function protects markdown links `[text](url)` from being broken by NIP-27
 * URL extraction. Markdown links are preserved and will be rendered correctly
 * when passed through renderMarkdown().
 */
export async function parseContent(content: string): Promise<ContentBlock[]> {
	const blocks: ContentBlock[] = [];

	// Step 1: Extract and protect markdown links `[text](url)`
	// We replace them with placeholders, then restore after NIP-27 parsing
	const markdownLinks: Array<{ text: string; url: string; fullMatch: string }> = [];
	let protectedContent = content.replace(
		/\[([^\]]+)\]\(([^)]+)\)/g,
		(match, linkText, url) => {
			markdownLinks.push({ text: linkText, url, fullMatch: match });
			return MARKDOWN_LINK_PLACEHOLDER;
		}
	);

	// Step 2: Use NIP-27 to parse the protected content
	const nip27Blocks = Array.from(nip27.parse(protectedContent));

	let linkIndex = 0;
	for (const block of nip27Blocks) {
		switch (block.type) {
			case 'text': {
				// Restore markdown links within text blocks
				const textWithLinks = restoreMarkdownLinks(block.text, markdownLinks, linkIndex);
				linkIndex = textWithLinks.newIndex;

				// Parse the restored text for code blocks, cashu, hashtags
				const textBlocks = await parseTextBlock(textWithLinks.text);
				blocks.push(...textBlocks);
				break;
			}
			case 'reference': {
				// Nostr reference (npub, nprofile, note, nevent, naddr)
				const refBlock = parseReferenceBlock(block.pointer);
				if (refBlock) {
					blocks.push(refBlock);
				}
				break;
			}
			case 'url': {
				// Skip URLs that were part of markdown links (they were replaced with placeholder)
				if (block.url !== MARKDOWN_LINK_PLACEHOLDER) {
					blocks.push({
						type: 'link',
						text: block.url,
						data: { href: block.url }
					});
				}
				break;
			}
			case 'image': {
				blocks.push({
					type: 'image',
					text: block.url,
					data: { src: block.url }
				});
				break;
			}
			case 'video': {
				blocks.push({
					type: 'video',
					text: block.url,
					data: { src: block.url }
				});
				break;
			}
			case 'audio': {
				// Treat audio as link for now
				blocks.push({
					type: 'link',
					text: block.url,
					data: { href: block.url }
				});
				break;
			}
			case 'relay': {
				// Treat relay as link for now
				blocks.push({
					type: 'link',
					text: block.url,
					data: { href: block.url }
				});
				break;
			}
		}
	}

	// Post-processing: group consecutive media into grids
	return groupMediaIntoGrids(blocks);
}

/**
 * Restore markdown links from placeholders in text
 */
function restoreMarkdownLinks(
	text: string,
	links: Array<{ text: string; url: string; fullMatch: string }>,
	startIndex: number
): { text: string; newIndex: number } {
	let result = text;
	let currentIndex = startIndex;
	let replacedCount = 0;

	// Replace each placeholder with the actual markdown link
	while (result.includes(MARKDOWN_LINK_PLACEHOLDER) && currentIndex < links.length) {
		const link = links[currentIndex];
		result = result.replace(MARKDOWN_LINK_PLACEHOLDER, link.fullMatch);
		currentIndex++;
		replacedCount++;
	}

	return { text: result, newIndex: currentIndex };
}

/**
 * Parse a text block for code blocks, cashu tokens, and hashtags
 */
async function parseTextBlock(text: string): Promise<ContentBlock[]> {
	const blocks: ContentBlock[] = [];

	// Define patterns to match within text
	const patterns = [
		{
			type: 'code' as const,
			regex: /```([\s\S]*?)```/g,
			processMatch: (match: RegExpExecArray) => ({
				type: 'code' as const,
				text: match[0],
				data: { code: match[1] }
			})
		},
		{
			type: 'cashu' as const,
			regex: /(cashuA[A-Za-z0-9_-]+)/g,
			processMatch: (match: RegExpExecArray) => ({
				type: 'cashu' as const,
				text: match[0],
				data: { token: match[0] }
			})
		},
		{
			type: 'hashtag' as const,
			// Match hashtags that are not part of a URL
			regex: /(?<![^\s"'(])(#[a-zA-Z0-9_]+)(?![a-zA-Z0-9_])/g,
			processMatch: (match: RegExpExecArray) => ({
				type: 'hashtag' as const,
				text: match[0],
				data: { tag: match[0].substring(1) }
			})
		}
	];

	// Find all matches with their positions
	const allMatches: Array<{
		start: number;
		end: number;
		block: ContentBlock;
	}> = [];

	for (const pattern of patterns) {
		let match: RegExpExecArray | null;
		pattern.regex.lastIndex = 0;

		while ((match = pattern.regex.exec(text)) !== null) {
			const start = match.index;
			const end = start + match[0].length;
			const block = pattern.processMatch(match);
			allMatches.push({ start, end, block });
		}
	}

	// Sort matches by start position
	allMatches.sort((a, b) => a.start - b.start);

	// Remove overlapping matches
	const filteredMatches: typeof allMatches = [];
	for (const match of allMatches) {
		const overlaps = filteredMatches.some(
			(existing) =>
				(match.start >= existing.start && match.start < existing.end) ||
				(match.end > existing.start && match.end <= existing.end) ||
				(match.start <= existing.start && match.end >= existing.end)
		);
		if (!overlaps) {
			filteredMatches.push(match);
		}
	}

	// Re-sort filtered matches
	filteredMatches.sort((a, b) => a.start - b.start);

	// Build the final result
	let lastIndex = 0;
	for (const { start, end, block } of filteredMatches) {
		// Add text before this match
		if (start > lastIndex) {
			blocks.push({
				type: 'text',
				text: text.substring(lastIndex, start)
			});
		}
		blocks.push(block);
		lastIndex = end;
	}

	// Add any remaining text
	if (lastIndex < text.length) {
		blocks.push({
			type: 'text',
			text: text.substring(lastIndex)
		});
	}

	// If no matches, return the whole text
	if (blocks.length === 0 && text) {
		blocks.push({ type: 'text', text });
	}

	return blocks;
}

/**
 * Parse a NIP-27 reference into a ContentBlock
 */
function parseReferenceBlock(
	pointer: ProfilePointer | AddressPointer | EventPointer
): ContentBlock | null {
	try {
		// Determine the type of reference
		if ('pubkey' in pointer && !('identifier' in pointer)) {
			// nprofile or npub
			const bech32 = nip19.npubEncode(pointer.pubkey);
			return {
				type: 'npub',
				text: `nostr:${bech32}`,
				data: {
					decoded: pointer,
					bech32
				}
			};
		} else if ('identifier' in pointer) {
			// naddr
			const addrPointer = pointer as AddressPointer;
			const bech32 = nip19.naddrEncode({
				kind: addrPointer.kind,
				pubkey: addrPointer.pubkey,
				identifier: addrPointer.identifier,
				relays: addrPointer.relays
			});
			return {
				type: 'naddr',
				text: `nostr:${bech32}`,
				data: {
					decoded: pointer,
					bech32
				}
			};
		} else if ('id' in pointer) {
			// nevent or note
			const eventPointer = pointer as EventPointer;
			const bech32 = nip19.noteEncode(eventPointer.id);
			return {
				type: 'note',
				text: `nostr:${bech32}`,
				data: {
					decoded: pointer,
					bech32
				}
			};
		}
	} catch (e) {
		console.error('Failed to parse reference:', e);
	}
	return null;
}

/**
 * Group consecutive media blocks into grids
 */
function groupMediaIntoGrids(blocks: ContentBlock[]): ContentBlock[] {
	const processedBlocks: ContentBlock[] = [];
	let mediaGroup: ContentBlock[] = [];

	for (let i = 0; i < blocks.length; i++) {
		const block = blocks[i];

		if (block.type === 'image' || block.type === 'video') {
			mediaGroup.push(block);
			continue;
		}

		// If this is whitespace between media, check what follows
		if (block.type === 'text' && /^\s+$/.test(block.text)) {
			if (
				mediaGroup.length > 0 &&
				i + 1 < blocks.length &&
				(blocks[i + 1].type === 'image' || blocks[i + 1].type === 'video')
			) {
				continue;
			}
		}

		// If we have collected media and the current block breaks the sequence
		if (mediaGroup.length > 0) {
			if (mediaGroup.length > 1) {
				processedBlocks.push({
					type: 'mediaGrid',
					text: mediaGroup.map((m) => m.text).join('\n'),
					data: {
						items: mediaGroup.map((m) => ({
							type: m.type,
							src: m.data?.src
						}))
					}
				});
			} else {
				processedBlocks.push(mediaGroup[0]);
			}
			mediaGroup = [];
		}

		processedBlocks.push(block);
	}

	// Don't forget any remaining media
	if (mediaGroup.length > 0) {
		if (mediaGroup.length > 1) {
			processedBlocks.push({
				type: 'mediaGrid',
				text: mediaGroup.map((m) => m.text).join('\n'),
				data: {
					items: mediaGroup.map((m) => ({
						type: m.type,
						src: m.data?.src
					}))
				}
			});
		} else {
			processedBlocks.push(mediaGroup[0]);
		}
	}

	return processedBlocks;
}

/**
 * Render markdown text to HTML using marked
 * This should be used for text blocks in articles (kind 30023)
 */
export function renderMarkdown(content: string): string {
	return marked.parse(content, {
		breaks: true, // Convert \n to <br>
		gfm: true, // GitHub Flavored Markdown
		async: false // Return string synchronously
	}) as string;
}
